runtime/sqlite_record_logger.hpp

```diff
@@ -176,14 +177,16 @@ class sqlite_record_logger : public record_logger {
 private:
 	sqlite_thread& get_sqlite_thread(const record& r) {
 		const record_header& rh = r.get_record_header();
-		auto result = registered_tables.find(rh.get_id());
-		if (result != registered_tables.cend()) {
-			return result->second;
-		} else {
-			const std::lock_guard lock{_m_registry_lock};
-			auto pair = registered_tables.try_emplace(rh.get_id(), rh);
-			return pair.first->second;
+		{
+			const std::shared_lock<std::shared_mutex> lock {_m_registry_lock};
+			auto result = registered_tables.find(rh.get_id());
+			if (result != registered_tables.cend()) {
+				return result->second;
+			}
 		}
+		const std::unique_lock<std::shared_mutex> lock{_m_registry_lock};
+		auto pair = registered_tables.try_emplace(rh.get_id(), rh);
+		return pair.first->second;
```

Comment:
Just to make sure, this locking is required to support the case where two different SB threads may try to log the same record type at the same time? Is this use case being exercised in our code at the moment?

Reply:
Yeah; All switchboard threads submit `switchboard_callback` records to the same table. Only one of them should do the SQL `CREATE` incantation.

The SQL `INSERT` incantation is already serialized because all producers put in the same queue, and only one thread dequeues from that, which is responsible for actually doing the insertion,